src/proposed/0180-fpu-switching.md

```diff
@@ -0,0 +1,378 @@
+
+  In this case, it would still be possible to emulate the fault-based lazy
+  scheme from user-level:
+
+  - the rare-FPU thread is set to FPU disabled
+  - when it does use the FPU, the user-level fault handler gets the
+    corresponding fault
+  - the fault handler enables the FPU on this thread and restarts it
+  - the thread could either stay FPU enabled or be FPU-disabled again after its
+    time slice or budget are up, depending on system policy.
+
+  This does have more overhead than the current in-kernel transparent fault
+  handling, but not massively more, and it allows full user-level control and
+  trade-offs. For instance, this could also be used for starting all threads as
+  FPU-disabled and enabling FPU for them on demand. That would yield low
+  amortised overhead without needing system analysis.
+
+## Rationale and alternatives
+
+There are multiple alternatives to the semi-lazy context switching scheme
+proposed here. The main classes are:
+
+- fully eager -- FPU context is always saved and restored
+- semi-lazy with early context saving -- FPU context is always saved when
+  switching away from an FPU-enabled thread and always restored when switching to
+  a FPU-enabled thread
+- fault-based lazy -- the current scheme in seL4
+
+This RFC has already argued that semi-lazy is preferable to the current scheme
+for multiple reasons (policy, performance, verification). Compared to eager
+switching, the proposal has the following advantages:
+
+- Fully eager represents another fixed kernel policy vs configurable policy.
+- Fully eager switching can be achieved in the proposed scheme by setting all
+  threads to FPU enabled. The only overhead this is incurs is one branch test
+  which checks the flag and always finds the FPU enabled.
+- Fully eager switching does not perform well on all work loads, for example a
+  common use case in the device driver framework is a chain of OS
+  producers/consumers that do not use FPU, called by an application that
+  potentially does use the FPU.
+- If the one-branch-test overhead for fully eager switching in the proposed
+  scheme is important for a specific workload, it would be possible to provide
+  an optimisation: have an always-eager compile-time config flag that statically
+  sets the `fpuDisabled` flag to `false`, which eliminates the branch. This
+  would disallow disabling the FPU in `seL4_TCB_Set_Flags`. We do not anticipate
+  this to be needed currently, but it is an option for the future if this turns
+  out to be a common case.
+
+The proposed scheme is slightly more complex than the semi-lazy scheme with
+early context save, and early saving would not need any special treatment on
+domain switch, but the targeted use case of OS threads without FPU called by an
+application thread with FPU would still need one FPU context save and restore,
+whereas the proposed scheme does not require any.
+
+
+## Prior art
```

Comment:
For the domain switching, the Arm-hyp version has the same confidentiality (infoflow) issue, *but* relevant aspects of VCPU switching are in the verified subset. Switching to NULL (None) VCPU on domain switch is a tweak we want to introduce in the future, but one that won't require an RFC. It would also require an invariant that the current owner of the VCPU is in the current domain.

So, this RFC would make the FPU switching like VCPU switching is currently, plus plugging the hole you pointed out, so we can re-use a bunch of the proof techniques. Does this help at all?

Reply:
Thanks @Xaphiosis , your explanations are helpful.